[OpenMP][Flang] Fix atomic operations on complex types

KrxGu opened the pull request:
Fixes #165184

## Problem
When using OpenMP atomic read/write on Fortran `complex(8)` (16 bytes) or `complex(16)` (32 bytes), only the real part was being stored/loaded. The imaginary part remained unchanged (typically 0).

## Root Cause
In `OMPIRBuilder::createAtomicRead()` and `createAtomicWrite()`, the size parameter for `__atomic_load`/`__atomic_store` was incorrectly computed from the pointer type instead of the pointee (element) type.

On 64-bit systems, this resulted in only 8 bytes being transferred regardless of the actual struct size.

## Solution
Changed both functions to use `XElemTy` (element type) instead of the pointer type when computing `LoadSize`. This ensures the full struct is transferred:
- `complex(4)`: 8 bytes ✓
- `complex(8)`: 16 bytes ✓ (was 8 before)
- `complex(16)`: 32 bytes ✓ (was 8 before)

## Tests
Added three regression tests:
- `atomic-write-complex.f90` - Verifies IR for atomic write
- `atomic-read-complex.f90` - Verifies IR for atomic read  
<img width="697" height="196" alt="image" src="https://github.com/user-attachments/assets/1256c925-c7e3-4634-bd4a-7f56a4d90384" />

- `issue-165184-atomic-complex8.f90` - Direct reproducer from the issue

kparzysz reviewed (approved):
LGTM

Commit:
[OpenMP][Flang] Fix atomic size for complex types

Fixes #165184

In OMPIRBuilder::createAtomicRead() and createAtomicWrite(), the size
parameter for __atomic_load/__atomic_store was incorrectly computed
from the pointer type instead of the pointee (element) type.

On 64-bit systems, this resulted in only 8 bytes being transferred
regardless of the actual struct size.

Changed both functions to use XElemTy (element type) instead of the
pointer type when computing LoadSize. This ensures the full struct
is transferred.

kiranchandramohan commented on flang/test/Lower/OpenMP/atomic-read-complex.f90:
```diff
@@ -0,0 +1,34 @@
+! Test lowering of atomic read to LLVM IR for complex types.
+! This is a regression test for issue #165184.
+
+! RUN: %flang_fc1 -emit-llvm -fopenmp -o - %s | FileCheck %s
```
We should try to avoid LLVM tests in the Lower directory.

KrxGu commented:
@kparzysz Commit message updated and squashed per review. CI all green!!

@kiranchandramohan Re: colocating tests with the fix in OpenMPIRBuilder, plan is:
- Add a gtest in `llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp` that builds atomic write/read on `{double,double}` and asserts the first arg to `__atomic_{load,store}` is 16.
- Add an MLIR translation check in `mlir/test/Target/LLVMIR/openmp-llvm.mlir` for `omp.atomic` lowering to `__atomic_*` with `i64 16`.

I can keep the Flang IR tests as frontend coverage, or drop them if you prefer only the two tests above. Please advise.

@kparzysz Does this test plan align with your expectations? and are we on the same page?

tblah reviewed (commented):
Thank you for the bug fix!

Commit:
[OpenMP][Flang] Fix atomic operations size for complex types

OpenMP atomic operations on complex(8) were using pointer size (8 bytes)
instead of the actual element size (16 bytes), causing only 8 bytes to
be stored/loaded instead of the full 16 bytes.

Fixed by using DL.getTypeStoreSize(XElemTy) instead of pointer type size
in createAtomicRead and createAtomicWrite for both atomic operations.

Added Flang IR tests to verify correct size generation for complex(4)
and complex(8) atomic operations, and unit test to verify struct types
with __atomic_store/__atomic_load.

Fixes #165184

KrxGu commented:
@kiranchandramohan ,  @kparzysz  - I've addressed the feedback and made the requested changes. The PR now has:

**Unit test** for atomic operations on struct types (test `OMPAtomicRWStructType` in `OpenMPIRBuilderTest.cpp`)
   - Verifies that `__atomic_store` and `__atomic_load` are called with the correct size (16 bytes) for `{double, double}` struct type
   - Confirms the fix works for struct types similar to complex(8)
   
The commit history has been squashed into a single clean commit with a descriptive message. Please review when you get a chance. CI is all green, Good to Merge!!

tblah reviewed (approved):
LGTM, thanks!
